lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +55,129 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    contextHandler = new class
+      @removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
+
+      @getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while getContext(editor.session, i - 1) == "equation"
+          i -= 1
+        start = i
+        while getContext(editor.session, i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      @getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(@removeRegex, "")
+
+      @getPopoverPosition: (row) -> {
+          top: "#{editor.renderer.textToScreenCoordinates(row + 2, 1).pageY}px"
+          left: "#{jqEditorContainer.position().left}px"
+        }
+
+      constructor: ->
+        [@curStart, @curEnd] = [null, null]
```

Comment:
`curStart` and `curEnd` are actually used only in the single call tree: initPopover + getEquationRange + getCurrentFormula. I suggest removing them and passing positions between these calls as arguments. This will make the code more understandable, because now when one looks at e.g. `getCurrentFormula` he needs to answer the question "do I understand why curstart and curEnd values at this moment are the actual ones?"